### PR TITLE
openapi spec for the wallet backend

### PR DESCRIPTION
**Pull request by gouthamp-stellar**
### What

The openapi spec for the wallet backend

### Why

For api documentation

### Known limitations

N/A

### Issue that this PR addresses

https://github.com/orgs/stellar/projects/58/views/2?pane=issue&itemId=85462088&issue=stellar%7Cwallet-backend%7C69

#### PR Structure

- [x] It is not possible to break this PR down into smaller PRs.
- [x] This PR does not mix refactoring changes with feature changes.
- [x] This PR's title starts with name of package that is most changed in the PR, or `all` if the changes are broad or impact many packages.

#### Thoroughness

- [x] This PR adds tests for the new functionality or fixes.
- [x] All updated queries have been tested (refer to [this](https://stackoverflow.com/a/29163465) check if the data set returned by the updated query is expected to be same as the original one).

#### Release

- [x] This is not a breaking change.
- [x] This is ready to be tested in development.
- [x] The new functionality is gated with a feature flag if this is not ready for production.

### CODE REVIEW

File: openapi/main.yaml [hunk 1]
```diff
@@ -0,0 +1,704 @@
...
+                          type: string
+                        destAmount:
+                          type: integer
+                        destAssetCode:
+                          type: string
+                        destAssetIssuer:
+                          type: string
+                        fromAddress:
+                          type: string
+                        memo:
```
Comment (CassioMG) [inline openapi/main.yaml:196]: should we add an optional `memoType` in addition to the `memo` value?

File: openapi/main.yaml [hunk 2]
```diff
@@ -0,0 +1,704 @@
...
+                      next:
+                        type: string
+                      prev:
+                        type: string
+                  payments:
+                    type: array
+                    description: list of payments 
+                    items:
+                      type: object
+                      properties:
```
Comment (CassioMG) [inline openapi/main.yaml:185]: should it cover some [path payment properties](https://developers.stellar.org/docs/learn/fundamentals/transactions/list-of-operations#payment) like the below?
- `path` (e.g. ["XLM", "BTC", "USDC"]
- `destinationMin`
- `sendMax
- and maybe a `kind` prop (e.g. PAYMENT | PATH_PAYMENT_STRICT_SEND | PATH_PAYMENT_STRICT_RECEIVE)

Also, I am assuming we are not supporting Anchor Transaction properties on those payment objects. Is that right?
E.g.: `anchor transaction id`,  `kind`, `status`, `more_info_url`, etc...

**Comment by JakeUrban** [comment]
@CassioMG lets hold off on discussions about changes we want to make to the API in this PR and focus on the spec itself for now. I do want to review the API's design holistically but lets do that separately.

File: openapi/main.yaml (hunk 2 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml:185]: All good points @CassioMG. And there's certainly other things that I think need to change in this API as well. I'm going to start a doc with all these recommendations

**Comment by CassioMG** [comment]
> @CassioMG lets hold off on discussions about changes we want to make to the API in this PR and focus on the spec itself for now. I do want to review the API's design holistically as a team but lets keep that discussion separate.

Sounds good to me

File: openapi/main.yaml [hunk 3]
```diff
@@ -0,0 +1,704 @@
...
+  termsOfService: https://stellar.org/terms-of-service
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+servers:
+  - url: ''
+tags:
+  - name: Account Registration
+    x-displayName: Account Registration
+    description: Clients of the Wallet Backed API can register and de-register stellar accounts whose payments they want to track
```
Comment (JakeUrban) [inline openapi/main.yaml:18]: So to confirm, registration is only relevant to the data returned from `/payments`, right? These accounts are irrelevant to the `tx` and `tss` endpoints?

File: openapi/main.yaml [hunk 4]
```diff
@@ -0,0 +1,704 @@
...
+                    properties:
+                      operations:
+                        type: array
+                        description: list of operations that make up the transaction
+                        items:
+                          type: string
+                          description: base64 operation xdr string
+                      timebounds:
+                        type: integer
+                        description: time in seconds after which this transaction will not be accepted by the network
```
Comment (JakeUrban) [inline openapi/main.yaml:480]: Is this the number of seconds _after the current time_ that the transaction expires, or the unix timestamp for when the transaction expires?

File: openapi/main.yaml [hunk 5]
```diff
@@ -0,0 +1,704 @@
...
+              required:
+                - transactions
+            example:
+              transactions:
+                - opearations:
+                  - "Y6MF7SMT2a2d6pt3i37Xx9"
+                  - "M8GA7SMT2a2d6pt3i37Xx8"
+                  timebounds: 300
+      responses:
+        '200':
```
Comment (JakeUrban) [inline openapi/main.yaml:493]: So 200 is returned if all transactions provided in the request were built successfully, and 400 is returned if at least 1 transaction couldn't be built?

File: openapi/main.yaml [hunk 6]
```diff
@@ -0,0 +1,704 @@
...
+                  type: string
+                  description: The url to which the final transaction results are to be sent
+                transactions:
+                  type: array
+                  items:
+                    type: string
+                    description: base64 encoded transaction xdr to submit to the stellar network
+                feebump:
+                  type: boolean
+                  description: boolean indicating if this transaction should be wrapped in a fee bump transaction
```
Comment (JakeUrban) [inline openapi/main.yaml:565]: Would it be equivalent for the client to submit transactions individually to `/tx/create-fee-bump`, and then submit all of them to this endpoint with the `feebump` parameter set to `false`?

File: openapi/main.yaml [hunk 7]
```diff
@@ -0,0 +1,704 @@
...
+                - transactions
+            example:
+              webhookurl: "www.sdp.stellar.com/webhook"
+              transactions:
+                - "Y6MF7SMT2a2d6pt3i37Xx9"
+                - "M8GA7SMT2a2d6pt3i37Xx8"
+              feebump: true
+      responses:
+        '200':
+          description: "Successful response with the transaction hashes of the transactions to be submitted to the network. Clients can use these hashes to query TSS for the status of their transaction"
```
Comment (JakeUrban) [inline openapi/main.yaml:577]: Does this endpoint submit all transactions provided in the request to RPC before returning? Or are they just queued for submission?

File: openapi/main.yaml [hunk 8]
```diff
@@ -0,0 +1,704 @@
...
+                properties:
+                  transactionhash:
+                    type: string
+                    description: "The hash of the transaction."
+                  transactionxdr:
+                    type: string
+                    description: "The base64-encoded xdr string of the transaction"
+                  status:
+                    type: string
+                    description: "The current status of the transaction"
```
Comment (JakeUrban) [inline openapi/main.yaml:652]: What are the values that can be returned here and their definitions?

File: openapi/main.yaml [hunk 9]
```diff
@@ -0,0 +1,704 @@
...
+      parameters:
+        - name: address
+          in: query
+          description: The stellar address to register with the wallet backend
+          required: true
+          schema:
+            type: string
+      responses:
+        '200':
+          description: OK
```
[the remark this answers is not in the document]
Comment (gouthamp-stellar) [reply openapi/main.yaml:42]: The API does not have a response body for the registration endpoints as of now

File: openapi/main.yaml [hunk 10]
```diff
@@ -0,0 +1,704 @@
...
+  /tx/create-fee-bump:
+    post:
+      tags:
+        - Transactions
+      summary: Create a fee bump transaction where the distribution account pays for the fee.
+      description:
+        Create a fee bump transaction where the distribution account pays for the transaction fee.
+      operationId: createFeeBumpAccount
+      requestBody:
+        description: "Request body for configuring address signers with weights and types."
```
[the remark this answers is not in the document]
Comment (gouthamp-stellar) [reply openapi/main.yaml]: Ha, yeap. Changed this to reflect the endpoint accurately

File: openapi/main.yaml (hunk 3 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml]: It actually also applies to the create-fee-bump-tx endpoint - ie: only transactions with stellar accounts that are registered with the wallet backend as the source can call the create-fee-bump-tx endpoint. We may need to revisit that logic though

File: openapi/main.yaml (hunk 4 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml]: Yeap, it is the number of seconds after the current time that the transaction expires. Changed this description to exactly that

File: openapi/main.yaml (hunk 5 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml:493]: That's right. It's all or nothing. Either all transactions are built successfully or none of them are, from the client's perspective

File: openapi/main.yaml (hunk 6 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml:565]: Yes thats right. Of course, that would add an additional endpoint for clients of TSS to call though before submitting them. I have to add though that in the /tx/create-fee-bump has some additional logic where it blocks certain operation types (which can be configured). Open to changing the api design around these endpoints though. I'm open to removing the feebump boolean for example

File: openapi/main.yaml (hunk 7 above)
Comment (gouthamp-stellar) [reply openapi/main.yaml:577]: Nope. Transactions are queued for submission. These transaction hashes are just receipts that the client can use to call the GET /tss/transactions/{txhash} endpoint

**Review by JakeUrban — APPROVED** [review]
LGTM. I'm sure we'll refer back to this PR and the comments made when we're reviewing the API design itself.

**Pull request merged**